### PR TITLE
Add OST stats

### LINKED ISSUE #53: Support for ost stats
`ost.OSS.ost.stats`:
```
ost.OSS.ost.stats=
snapshot_time             1688128253.497763049 secs.nsecs
req_waittime              18419628 samples [usec] 2 40983 305482965 25043535105
req_qdepth                18419628 samples [reqs] 0 34 99937 130635
req_active                18419628 samples [reqs] 1 36 69585063 634492353
req_timeout               18419628 samples [sec] 1 15 276294334 4144414654
reqbuf_avail              38185151 samples [bufs] 60 64 2438170078 155685175822
ldlm_glimpse_enqueue      9257180 samples [reqs] 1 1 9257180 9257180
ldlm_extent_enqueue       19856 samples [reqs] 1 1 19856 19856
ost_create                144904 samples [usec] 6 16594 98795730 85661707326
ost_destroy               8988941 samples [usec] 89 173579 5160119682 8184502010174
ost_get_info              8 samples [usec] 540 3603 10971 28145019
ost_connect               341 samples [usec] 21 903 24182 2818080
ost_disconnect            331 samples [usec] 23 524 39358 7068516
ost_sync                  4510 samples [usec] 3 10945 997271 2117171965
ost_set_info              28 samples [usec] 9 34 606 14594
obd_ping                  3529 samples [usec] 3 12431 60722 155336592
```

`ost.OSS.ost_io.stats`:
```
ost.OSS.ost_io.stats=
snapshot_time             1688128269.170769339 secs.nsecs
req_waittime              3398592545 samples [usec] 2 585517 95316362073 32500246129015
req_qdepth                3398592545 samples [reqs] 0 53 90676247 112259319
req_active                3398592545 samples [reqs] 1 82 55496806665 1427593461517
req_timeout               3398592545 samples [sec] 15 15 50978888175 764683322625
reqbuf_avail              7234158916 samples [bufs] 55 64 461878663298 29490702443182
ost_read                  2447557926 samples [usec] 23 138321 1871024223288 4497819893384848
ost_write                 951033049 samples [usec] 59 1247713 2749050524782 100048363896296658
ost_punch                 1515 samples [usec] 16 4883 63967 29511205
```

`ost.OSS.ost_create.stats`:
```
ost.OSS.ost_create.stats=
snapshot_time             1688128286.100343697 secs.nsecs
req_waittime              1109111 samples [usec] 6 65076 70990198 730579977056
req_qdepth                1109111 samples [reqs] 0 1 580 580
req_active                1109111 samples [reqs] 1 2 1224421 1455041
req_timeout               1109111 samples [sec] 10 15 16636655 249549725
reqbuf_avail              2307926 samples [bufs] 63 64 147667189 9448175371
ost_statfs                1109111 samples [usec] 3 80581 31692235 186522121515
```

### PR DESCRIPTION
**Pull request by RDruon**
Fix #53

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -496,6 +496,7 @@ pub enum TargetStats {
     ThreadsMax(TargetStat<u64>),
     ThreadsStarted(TargetStat<u64>),
     RecoveryStatus(TargetStat<RecoveryStatus>),
+    OstStat(TargetStat<Vec<Stat>>),
```
Comment (jgrund) [inline src/types.rs:499]: This is redundant with the already existing `Stats` variant

File: src/oss/ost_parser.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2021 DDN. All rights reserved.
```
Comment (jgrund) [inline src/oss/ost_parser.rs:1]: 2023

File: src/stats_parser.rs [hunk 3]
```diff
@@ -26,7 +26,7 @@ where
     I::Error: ParseError<I::Token, I::Range, I::Position>,
 {
     (
-        not_words(&["obdfilter", "mgs", "mdt", LDLM]).skip(spaces()),
+        not_words(&["obdfilter", "mgs", "mdt", LDLM, "ost"]).skip(spaces()),
```
Comment (jgrund) [inline src/stats_parser.rs:29]: Should be const

File: src/oss/ost_parser.rs
Comment (jgrund) [inline src/oss/ost_parser.rs:1]: It seems weird to have an oss parser and an ost parser. Just one module is needed for both.

File: src/oss/ost_parser.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+use crate::{
+    base_parsers::{param, period, target},
+    stats_parser::stats,
+    types::{Param, Record, Stat, Target, TargetStat, TargetStats, TargetVariant},
+};
+use combine::{attempt, choice, error::ParseError, parser::char::string, stream::Stream, Parser};
+
+pub(crate) const OST: &str = "ost";
+
```
Comment (jgrund) [inline src/oss/ost_parser.rs:13]:
```suggestion
```

File: src/oss/ost_parser.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+/// Parses the name of a target
+fn param_non_final<I>(x: &'static str) -> impl Parser<I, Output = Param>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    attempt(string(x).skip(period()))
+        .map(|x| Param(x.to_string()))
+        .message("while getting param non")
```
Comment (jgrund) [inline src/oss/ost_parser.rs:49]: Message doesn't make sense

File: src/oss/ost_parser.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        .collect()
+}
+
+/// Parses the name of a target
+fn target_name<I>() -> impl Parser<I, Output = Target>
+where
+    I: Stream<Token = char>,
+    I::Error: ParseError<I::Token, I::Range, I::Position>,
+{
+    (string("ost").skip(period()), target().skip(period()))
```
Comment (jgrund) [inline src/oss/ost_parser.rs:36]: const

File: src/oss/ost_parser.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+/// Takes OBD_STATS and produces a list of params for
+/// consumption in proper ltcl get_param format.
+pub(crate) fn params() -> Vec<String> {
+    OST_STATS
+        .iter()
+        .map(|x| format!("ost.OSS.{x}.stats"))
+        .collect()
+}
+
+/// Parses the name of a target
```
Comment (jgrund) [inline src/oss/ost_parser.rs:30]: fn name and comment don't agree with actual generality of code.

File: src/fixtures/valid/valid.txt [hunk 8]
```diff
@@ -276,3 +276,55 @@ req_active                1 samples [reqs] 1 1 1 1
 req_timeout               1 samples [secs] 15 15 15 225
 reqbuf_avail              3 samples [bufs] 1 1 3 3
 ldlm_bl_callback          1 samples [usecs] 16 16 16 256
+ost.OSS.ost.stats=
```
Comment (jgrund) [inline src/fixtures/valid/valid.txt:279]: This placement is invalid. stats outputs matches how they are specified in lctl

File: src/oss/ost_parser.rs
Comment (RDruon) [reply src/oss/ost_parser.rs:1]: OSS have at least 2 endpoint for their related metrics `obdfilter.*` and `ost.*`.
I thought that `oss_parser` was the top module that merge the `obdfilter.*` parser and the `ost.*` parser.

Let me know if you want me to move the code from `ost_parser.rs` to `oss_parser.rs`

File: src/oss/ost_parser.rs (hunk 7 above)
Comment (jgrund) [reply src/oss/ost_parser.rs]: Removed comment

File: src/oss/ost_parser.rs
Comment (jgrund) [reply src/oss/ost_parser.rs:1]: parser added in this PR is OSS specific. It encodes 0 information about OST. oss_parser makes more sense.

I pushed a new patch that addresses this.

**Review by jgrund — APPROVED** [review]

**Pull request merged**